### PR TITLE
feat(embodiment): d4rl offline iql with fsdp strategy

### PR DESCRIPTION
**Pull request by Dps799**
# D4RL_OFFLINE_IQL 实现报告

## 1. 背景与目标
这次PR主要修改是落地了D4RL离线IQL的训练链路。核心目标是在不改动IQL本身数学逻辑的情况下接入RLinf现有的Worker/Runner体系，而且尽量和SAC的工程写法保持一致，减少后续维护的成本。IQL原始的论文实现为JAX + Flax + Optax，现在实现为PyTorch+RLinf(EmbodiedIQLFSDPPolicy + D4RLOfflineRunner)。

核心事情三件：
- 把IQL的训练、评估、日志记录、断点续训这些完整流程在RLinf里跑通
- 严格遵循IQL原有的更新顺序（先更Value → 再更Actor → 接着更Critic → 最后软更新Target）
- 模块组织方式尽量和现有代码对齐，方便后续扩展和维护

## 2. 总体设计
- 整体架构还是沿用Actor + Double Critic + Value + Target Critic这四件套，针对离线学习的场景只做了最必要的改造，没有额外加冗余逻辑：
- 完全砍掉了在线rollout/replay采样的依赖，毕竟是离线训练，只保留了一个worker只保留了一个actor worker.
- `run_training()`每一步只从D4RL数据集里采样一个batch

单步训练的流程可以拆成这几步：
1. 从`Dataset.sample(batch_size)`读取离线样本
2. 用target critic算出的`min(Q1, Q2)`来更新V(s)（也就是expectile loss）
3. 按照`exp((Q - V) * temperature)`这个权重更新策略的对数概率
4. 用`r + discount * mask * V(s')`这个目标值更新双Q网络（MSE损失）
5. 用tau对target critic做软更新

## 3. 数学定义（和代码一一对应）
这里把核心的数学公式列出来，和代码里的实现完全对齐，方便核对：

#### 3.1 Value 更新（Expectile Regression）
- 误差：`d = Q_target(s, a) - V(s)`
- 损失：`L_v = E[w(d) * d^2]`
- 权重（d > 0）：`w(d) = expectile`
- 权重（d <= 0）：`w(d) = 1 - expectile`

#### 3.2 Actor 更新（AWR 风格加权）
- 优势：`A(s, a) = Q_target(s, a) - V(s)`
- 权重：`w = exp(A * temperature)`
- 数值稳定：`w = min(w, 100)`
- 损失：`L_actor = -E[w * log pi(a|s)]`

#### 3.3 Critic 更新（TD 误差）
- 目标：`y = r + gamma * m * V(s_next)`
- 损失：`L_q = E[(Q1 - y)^2 + (Q2 - y)^2]`

### 4.1 Worker：`rlinf/workers/actor/fsdp_iql_policy_worker.py`
这是IQL训练的核心文件，所有训练相关的逻辑都集中在这：模型搭建、参数更新、评估、存档这些全在这处理。

主要实现的功能点：
- `setup_iql_components`：初始化D4RL环境、数据集、随机种子、计算设备和超参
- `setup_model_and_optimizer`：构建actor/critic/value/target这几套网络，以及对应的优化器
- `forward_value / forward_actor / forward_critic`：把三个模块的更新逻辑拆分开，更清晰
- `update_one_epoch`：把上面三个更新步骤串起来，最后再做target的软更新
- `run_training`：单步训练 + 定期评估 + 指标汇总
- `save_checkpoint / load_checkpoint`：
  - actor / critic / value 分模块通过`_strategy`接口独立存/读（避免参数映射混淆）
  - IQL特有的状态（`target_critic_q1q2.pt`、`state.pt`）单独处理，避免和其他模块混在一起

另外，为了和SAC的代码风格对齐，也做了这些细节：
- `setup_model_and_optimizer`、`build_lr_schedulers`、`save/load_checkpoint`这些函数的结构完全照着SAC来
- target模型的存/取用`_strategy.get_model_state_dict` / `_strategy.load_model_with_state_dict`
- 注释和变量命名都尽量贴近`fsdp_sac_policy_worker.py`的风格

### 4.2 Runner：`rlinf/runners/d4rl_offline_runner.py`
负责离线训练的主循环，还有日志、存档的调度工作。

核心能力：
- `init_workers`：启动actor进程，支持从指定的`resume_dir`恢复训练
- `run`：每一步调用`actor.run_training()`，同时记录耗时、训练指标、评估结果
- `_save_checkpoint`：按照配置的`save_interval`定期保存checkpoint
- 用`log_queue + log_thread`做异步终端打印，避免打印日志阻塞训练流程

### 4.3 Dataset：`rlinf/data/datasets/d4rl_offline.py`
专门处理D4RL数据的加载和评估的封装，让上层调用更统一。

主要做的事：
- 加载`qlearning_dataset`，并把数据整理成统一的字段：`observations/actions/rewards/masks/next_observations`
- 对不同环境做适配处理（比如MuJoCo的数据归一化、AntMaze的奖励偏移）
- 提供`evaluate_policy`统一的评估接口，和其他模块对齐

### 4.4 Policy：`rlinf/models/embodiment/mlp_policy/mlp_policy.py`
IQL和SAC共用的MLP策略实现，这里主要是给IQL加了适配逻辑。

关键改动：
- `mlp_policy/__init__.py` 的 `get_model` 接口保持原生签名，新增能力仅通过 `cfg.iql_config` 透传，不引入额外公开参数
- 加了`ForwardType.IQL`的路由，通过`iql_forward`统一处理actor/critic/value分支
- IQL的初始化逻辑封装在`_init_iql`里，不会影响原来SAC的默认逻辑
- `forward()`函数不再自动推断是不是IQL模式，改成调用方显式传`forward_type=ForwardType.IQL`，更清晰

### 4.5 参数文件（Config）
IQL离线训练目前使用以下参数文件：

- MuJoCo：`RLinf_merge/examples/embodiment/config/d4rl_offline_mujoco.yaml`
- AntMaze：`RLinf_merge/examples/embodiment/config/d4rl_offline_antmaze.yaml`
- Kitchen/Adroit：`RLinf_merge/examples/embodiment/config/d4rl_offline_kitchen_adroit.yaml`

统一入口脚本为：

- `RLinf_merge/examples/embodiment/run_d4rl_offline.sh`

说明：

- 上述三个配置文件分别对应不同数据集族，算法更新流程一致，仅环境名与部分超参不同。
- 断点续训统一通过配置项`runner.resume_dir`指定`.../checkpoints/global_step_xxx`目录。

### 5.0 环境安装（D4RL离线）
先安装离线IQL所需环境依赖：
```bash
cd ./RLinf
bash requirements/install.sh embodied --env d4rl_offline
```

### 5.1 单任务启动
直接跑这个脚本就行，指定预设的任务名：
```bash
cd ./RLinf
./examples/embodiment/run_d4rl_offline.sh
```

### 5.2 指定环境启动
指定具体的环境名，脚本会自动匹配对应的预设配置：
```bash
./examples/embodiment/run_d4rl_offline.sh halfcheetah-medium-v2
./examples/embodiment/run_d4rl_offline.sh antmaze-medium-diverse-v0
```

### 5.3 断点续训
只需要在配置里加一行，指定恢复的目录就行：
```
runner.resume_dir=<.../checkpoints/global_step_xxx>
```
或者在bash中加入后缀
例：保存模型：
```bash
./examples/embodiment/run_d4rl_offline.sh halfcheetah-medium-v2 runner.save_interval=10000
```
加载模型：
```bash
./examples/embodiment/run_d4rl_offline.sh halfcheetah-medium-v2 max_steps=10000 runner.resume_dir="./RLinf/logs/20260303-15:08:40-iql_halfcheetah-medium-v2/checkpoints/global_step_5000"
```

Runner会自动做两件事：
1) 分别加载 actor / critic / value 的checkpoint
2) 加载IQL特有的`iql_components/target_critic_q1q2.pt`和`state.pt`文件

### 5.4 YAML 参数说明（`examples/embodiment/config/d4rl_offline_*.yaml`）
下面对配置文件中的主要参数做统一说明，便于调参与复现实验。

#### 运行与日志相关
- `cluster.num_nodes`：集群节点数。单机实验通常为 `1`。
- `cluster.component_placement.actor`：actor 进程放置策略，`0-0` 表示单卡单进程。
- `env.group_name`、`rollout.group_name`、`rollout.pipeline_stage_num`：`EmbodiedFSDPActor`初始化所需字段。
- `actor.group_name`：actor worker group 名称，用于调度系统内部通信。
- `actor.model`、`actor.fsdp_config`、`actor.training_backend`：FSDP初始化与`_strategy`创建所需配置。
- `runner.max_epochs`：训练总轮次。当前按离线单步更新语义，通常与 `max_steps` 绑定。
- `runner.save_interval`：checkpoint 保存间隔（步）。默认设为 `${eval_interval}`。
- `runner.resume_dir`：断点续训目录，指向 `.../checkpoints/global_step_xxx`。
- `runner.logger.log_path`：实验日志输出根目录。
- `runner.logger.experiment_name`：实验名，会影响日志与checkpoint目录名。

#### 数据与评估相关
- `env_name`：D4RL 环境名（如 `halfcheetah-medium-v2`、`antmaze-medium-diverse-v0`）。
- `seed`：随机种子（环境、numpy、torch）。
- `batch_size`：每步离线采样的批大小。
- `eval_interval`：评估间隔（步）。
- `eval_episodes`：每次评估跑多少个episode。
- `max_steps`：训练最大步数。
- `tqdm`：是否显示进度条。

#### IQL 算法超参数（`algorithm.*`）
- `actor_lr`：策略网络学习率。
- `value_lr`：价值网络学习率。
- `critic_lr`：双Q网络学习率。
- `hidden_dims`：MLP 隐层维度，默认 `[256, 256]`。
- `discount`：折扣因子 `gamma`。
- `tau`：target critic 软更新系数。
- `expectile`：expectile regression 系数，决定 `V` 回归偏向。
- `temperature`：AWR 权重温度（`exp((Q-V)*temperature)`）。
- `dropout_rate`：IQL actor 的 dropout，比方 `kitchen/adroit` 常设为 `0.1`。
- `opt_decay_schedule`：学习率衰减策略（当前使用 `cosine`）。

#### 三类预设配置差异
- `d4rl_offline_mujoco.yaml`：常规 MuJoCo locomotion 任务，参数与原始论文相同。
- `d4rl_offline_antmaze.yaml`：AntMaze 稀疏奖励任务，参数与原始论文相同。
- `d4rl_offline_kitchen_adroit.yaml`：Kitchen/Adroit 任务，参数与原始论文相同。

### 6.1 TensorBoard指标
所有指标都按分类整理好了，方便看：
- 训练指标：`train/*`
  - 损失类：`train/critic_loss`、`train/value_loss`、`train/actor_loss`
  - 数值类：`train/q1`、`train/q2`、`train/v`
  - 优势值：`train/adv_mean`、`train/adv_std`
- 评估指标：`evaluation/*`
- 时间指标：`time/*`

### 6.2 设计约束
- 只保留IQL核心的训练指标，不额外加无关的损失项，避免日志冗余
- `global_step`由runner统一维护，再下发给actor，保证全局步数一致
- D4RL离线路径依赖 `gym==0.23.1`（与 d4rl 版本约束保持兼容）

### 7.1 资源占用与论文基准
 内存占用情况
 
IQL论文基准 

断点复训

### 7.2 任务曲线对比
下面每个任务都是左侧是本次实现的曲线，右侧是参考的复现曲线，seed=42，由于两套代码框架不同，所以同一种子的表现可能存在部分差距：

#### halfcheetah-medium-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### halfcheetah-medium-replay-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### halfcheetah-medium-expert-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### hopper-medium-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### hopper-medium-replay-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### hopper-medium-expert-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### walker2d-medium-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### walker2d-medium-replay-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### walker2d-medium-expert-v2
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### antmaze-umaze-v0
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### antmaze-umaze-diverse-v0
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### antmaze-medium-play-v0
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### antmaze-medium-diverse-v0
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### antmaze-large-play-v0
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

#### antmaze-large-diverse-v0
| 本实现曲线 | 对比复现曲线 |
|---|---|
|  |  |

### CODE REVIEW

**Comment by Dps799** [comment]
tensorboard results url:https://huggingface.co/datasets/dps799/RLinf_iql_tensorboard
